File: NameComparator/NameComparator.py
from dataclasses import dataclass
from typing import NamedTuple
from json import loads as json_loads
from importlib.resources import files

from NameComparator.src.clean import clean_name, clean_names_by_comparison, clean_ipa
from NameComparator.src.nicknames import remove_nicknames
from NameComparator.src.insights import is_worth_continuing, either_name_too_short
from NameComparator.src.comparisons import compare_spelling, pronunciation_comparison
from NameComparator.src.modify import modify_names_together, modify_ipas_by_comparison
from NameComparator.src.ipa import get_ipa
from NameComparator.src.uniqueness import score_uniqueness
from NameComparator.src.uniqueness import FrequencyData

unparsed_usa_to_1950_surnames = files('data').joinpath('frequency/surnamesUsaTo1950.json').read_text()
unparsed_usa_to_1950_first_names = files('data').joinpath('frequency/firstNamesUsaTo1950.json').read_text()

class Attempt(NamedTuple):
    """Represents an attempt at name comparison (often used for debugging).

    Attributes:
        name_one: The version of the first name to be used in this attempt
        name_two: The version of the second name to be used in this attempt
        word_combos: A list of tuples describing the word matchups and quality
        score_of_attempt: The score associated with the percent confidence 
            returned from this attempt
    """
    name_one: str
    name_two: str
    word_combos: list[tuple[str, str, int]]
    score_of_attempt: float

@dataclass
class ResultsOfNameComparison:
    """Represents the results of a name comparison.

    Attributes:
        name_one: The original name_one
        name_two: The original name_two
        match: Whether or not the names are a match. Defaults to False
        uniqueness: How unique the names were in comparison to the chosen population. Defaults to 0.0
        too_short: Whether or not either of the names are one word or less. Defaults to True
        attempt_one: Debugging data about the first attempt to compare the names. Defaults to None
        attempt_two: Debugging data about the second attempt to compare the names. Defaults to None
        attempt_three: Debugging data about the third attempt to compare the names. Defaults to None
        attempt_four: Debugging data about the fourth attempt to compare the names. Defaults to None
        most_recent_attempt_score: The percent confidence score associated with the most recent attempt
            that was made while comparing the names
        average_score_of_combined_attempts: The average percent confidence score from all of the
            attempts that were made while comparing the names
    """
    name_one: str
    name_two: str
    match: bool = False
    uniqueness: float = 0.0
    too_short: bool = True
    attempt_one: Attempt | None = None
    attempt_two: Attempt | None = None
    attempt_three: Attempt | None = None
    attempt_four: Attempt | None = None
    most_recent_attempt_score: float = 0
    average_score_of_combined_attempts: float = 0

def compare_two_names(name_one:str, name_two:str, frequency_data:FrequencyData|None = None) -> ResultsOfNameComparison:
    """Compares two names to identify whether or not they are a match.

    Args:
        name_one: The first name to compare
        name_two: The second name to compare
        frequency_data: The first name and surname frequencies in a chosen 
            population - Defaults to None

    Returns:
        The data gleaned from the comparison: whether or not they are a match, 
        whether or not one or both names is too generic, whether or not one or 
        both names is too short, and the attempt data for each different 
        comparison method used
    """        
    # Deal with the optional frequency_data argument
    if not frequency_data:
        frequency_data = FrequencyData(json_loads(unparsed_usa_to_1950_first_names), json_loads(unparsed_usa_to_1950_surnames))

    # Data validation
    if not isinstance(name_one, str) or not isinstance(name_two, str):
        raise TypeError(f'name_one was {type(name_one)}. Must be a str. name_two was {type(name_two)}. Must be a str.')
    if not isinstance(frequency_data, FrequencyData):
        raise TypeError(f'frequency_data was the type {type(frequency_data)}. Must be a FrequencyData object.')

    # Create the return object to edit later
    results = ResultsOfNameComparison(name_one=name_one, name_two=name_two)

    # Clean the names
    name_one = clean_name(name_one)
    name_two = clean_name(name_two)
    name_one, name_two = clean_names_by_comparison(name_one, name_two)

    # Deal with names that are too short
    results.too_short = either_name_too_short(name_one, name_two)
    if not name_one:
        name_one = '_'
    if not name_two:
        name_two = '_'
    if (name_one == '_') or (name_two == '_'):
        return results
    
    # Find the uniqueness of this name matchup (ie. hopefully not 'John Smith' and 'J Smith')
    results.uniqueness = score_uniqueness(name_one, name_two, frequency_data)

    # Remove nicknames before the actual comparison
    name_one, name_two = remove_nicknames(name_one, name_two)

    # 1st attempt: Checks if names are a match according to string comparison alone
    match, word_combos, attempt_one_score = compare_spelling(name_one, name_two)
    results.attempt_one = Attempt(name_one, name_two, word_combos, attempt_one_score)
    if match:
        results.match = True
        results.most_recent_attempt_score = attempt_one_score
        results.average_score_of_combined_attempts = attempt_one_score
        return results

    # Failed first attempt. Check if names are even worth continuing
    if is_worth_continuing(name_one, name_two) is False:
        return results

    # 2nd attempt: Modify names via spelling rules, then check again if match according to string comparison
    modified_name_one, modified_name_two = modify_names_together(name_one, name_two)
    match, word_combos, attempt_two_score = compare_spelling(modified_name_one, modified_name_two)
    results.attempt_two = Attempt(modified_name_one, modified_name_two, word_combos, attempt_two_score)
    if match:
        results.match = True
        results.most_recent_attempt_score = attempt_two_score
        results.average_score_of_combined_attempts = ((attempt_two_score + attempt_one_score) / 2)
        return results
        
    # 3rd attempt: Checks if modified names are a match according to pronunciation
    ipa_of_modified_name_one = clean_ipa(get_ipa(modified_name_one))
    ipa_of_modified_name_two = clean_ipa(get_ipa(modified_name_two))
    ipa_of_modified_name_one, ipa_of_modified_name_two = modify_ipas_by_comparison(ipa_of_modified_name_one, ipa_of_modified_name_two)
    match, word_combos, attempt_three_score = pronunciation_comparison(ipa_of_modified_name_one, ipa_of_modified_name_two, modified_name_one, modified_name_two)
    results.attempt_three = Attempt(ipa_of_modified_name_one, ipa_of_modified_name_two, word_combos, attempt_three_score)
    if match:
        results.match = True
        results.most_recent_attempt_score = attempt_three_score
        results.average_score_of_combined_attempts = ((attempt_three_score + attempt_two_score + attempt_one_score) / 3)
        return results

    # 4th attempt: Check if original names are a match according to pronunciation
    ipa_of_name_one = clean_ipa(get_ipa(name_one))
    ipa_of_name_two = clean_ipa(get_ipa(name_two))
    ipa_of_name_one, ipa_of_name_two = modify_ipas_by_comparison(ipa_of_name_one, ipa_of_name_two)
    match, word_combos, attempt_four_score = pronunciation_comparison(ipa_of_name_one, ipa_of_name_two, name_one, name_two)
    results.attempt_four = Attempt(ipa_of_name_one, ipa_of_name_two, word_combos, attempt_four_score)
    if match:
        results.match = True
        results.most_recent_attempt_score = attempt_four_score
        results.average_score_of_combined_attempts = ((attempt_four_score + attempt_three_score + attempt_two_score + attempt_one_score) / 4)
    return results

File: NameComparatorTS/nameComparator.ts
import fs from 'fs';
import path from 'path';
import { cleanName, cleanNamesByComparison, cleanIpa } from './src/clean.js';
import { removeNicknames } from './src/nicknames.js';
import { isWorthContinuing, eitherNameTooShort } from './src/insights.js';
import { compareSpelling, pronunciationComparison } from './src/comparisons.js';
import { modifyNamesTogether, modifyIpasByComparison } from './src/modify.js';
import { getIpa } from './src/ipa.js';
import { scoreUniqueness, FrequencyData } from './src/uniqueness.js';

// This is required to make sure that it reads in the characters correctly
import usaTo1950SurnamesUnparsed from '../data/frequency/surnamesUsaTo1950.json' with {type: 'json'};
const usaTo1950Surnames = usaTo1950SurnamesUnparsed as Record<string, number>;

import usaTo1950FirstNamesUnparsed from '../data/frequency/firstNamesUsaTo1950.json' with  {type: 'json'};
const usaTo1950FirstNames = usaTo1950FirstNamesUnparsed as Record<string, number>;

// Read the penalty variable from a file
import values from '../data/variablesForComparisons.json' with { type: "json"};
const { maxScore, 
        guaranteedPassingScore,
        conditionallyPassingScore,
        furtherChecksNeededScore,
        guaranteedFailScore,
        fuzzyComparisonWeight,
        consonantComparisonWeight,
        numberOfValidCombosToSkipFurtherChecks,
        lengthNeededForConditionallyPassingPronunciationComparison,
        penaltyForMismatchedPrefixes
  } = values;

/**
 * Represents an attempt at name comparison (often used for debugging).
 * 
 * @property {string} nameOne - The version of the first name to be used in this attempt
 * @property {string} nameTwo - The version of the second name to be used in this attempt
 * @property {[string, string, number][]} wordCombos - A list of lists describing the word matchups and quality
 * @property {number} scoreOfAttempt - The score associated with the percent confidence returned from this attempt
 */
export class Attempt {
  constructor(
    public nameOne: string,
    public nameTwo: string,
    public wordCombos: [string, string, number][],
    public scoreOfAttempt: number
  ) {}
}

/**
 * Represents the results of a name comparison.
 * 
 * @property {string} nameOne - The original nameOne
 * @property {string} nameTwo - The original nameTwo
 * @property {boolean} match - Whether or not the names are a match. Defaults to false
 * @property {number} uniqueness - How unique the names were in comparison to the chosen population. Defaults to 0.0
 * @property {boolean} tooShort - Whether or not either of the names are one word or less. Defaults to true
 * @property {Attempt | null} attemptOne - Debugging data about the first attempt to compare the names. Defaults to null
 * @property {Attempt | null} attemptTwo - Debugging data about the second attempt to compare the names. Defaults to null
 * @property {Attempt | null} attemptThree - Debugging data about the third attempt to compare the names. Defaults to null
 * @property {Attempt | null} attemptFour - Debugging data about the fourth attempt to compare the names. Defaults to null
 * @property {number} mostRecentAttemptScore - The percent confidence score associated with the most recent attempt that 
 *                                             was made while comparing the names
 * @property {number} averageScoreOfCombinedAttempts - The average percent confidence score from all of the attempts that 
 *                                                     were made while comparing the names
 * @property {boolean} notWorthContinuing - Whether or not NameComparator determined if the name was not worth
 *                                          running more checks on after the first check
 */
export class ResultsOfNameComparison {
  constructor(
    public nameOne: string,
    public nameTwo: string,
    public match: boolean = false,
    public uniqueness: number = 0.0,
    public tooShort: boolean = true,
    public attemptOne: Attempt | null = null,
    public attemptTwo: Attempt | null = null,
    public attemptThree: Attempt | null = null,
    public attemptFour: Attempt | null = null,
    public mostRecentAttemptScore: number = 0.0,
    public averageScoreOfCombinedAttempts: number = 0.0,
    public notWorthContinuing: boolean = false
  ) {}
}

/**
 * Compares two names to identify whether or not they are a match.
 * 
 * @param nameOne - The first name to compare
 * @param nameTwo - The second name to compare
 * @param frequencyData - The first name and surname frequencies in 
 *                        a chosen population - Defaults to None
 * 
 * @returns The data gleaned from the comparison: whether or not they are a match, 
 *          whether or not one or both names is too generic, whether or not one or 
 *          both names is too short, and the attempt data for each different 
 *          comparison method used
 */
export function compareTwoNames(nameOne: string, nameTwo: string, frequencyData: FrequencyData | null = null): ResultsOfNameComparison {

  var appliedPenalty: number = 0;
  var shouldApplyPenalty = false;

  // Deal with the optional frequencyData argument
  if (!frequencyData) {
    frequencyData = new FrequencyData(usaTo1950FirstNames, usaTo1950Surnames);
  }

  // Data validation
  if (typeof nameOne !== 'string' || typeof nameTwo !== 'string') {
    throw new TypeError(`nameOne was ${typeof nameOne}. Must be string. nameTwo was ${typeof nameTwo}. Must be a string.`);
  }
  if (!(frequencyData instanceof FrequencyData)) {
    throw new TypeError(`frequencyData was the type ${typeof frequencyData}. Must be a FrequencyData boject.`);
  }

  // Create the return object to edit later
  let results = new ResultsOfNameComparison(nameOne, nameTwo);

  // Clean the names
  console.error(`Cleaning nameOne in TypeScript. Name before cleaning: ${nameOne}`);
  nameOne = cleanName(nameOne);
  console.error(`Cleaning nameTwo in TypeScript. Name before cleaning: ${nameTwo}`);
  nameTwo = cleanName(nameTwo);
  console.error("Cleaning names by comparison in TypeScript");
  [nameOne, nameTwo, shouldApplyPenalty] = cleanNamesByComparison(nameOne, nameTwo);
  console.error(`Names after cleaning in TypeScript - nameOne: ${nameOne}  nameTwo: ${nameTwo}`);

  if (shouldApplyPenalty === true){
    appliedPenalty = penaltyForMismatchedPrefixes
  };

  // Deal with names that are too short
  console.error("Determining if either name is too short in TypeScript");
  results.tooShort = eitherNameTooShort(nameOne, nameTwo);
  if (!nameOne) {
    nameOne = '_'
  };
  if (!nameTwo) {
    nameTwo = '_'
  };
  if (nameOne === '_' || nameTwo === '_'){
    return results;
  } 

  // Find the uniqueness of this name matchup (ie. hopefully not 'John Smith' and 'J Smith')
  console.error("Scoring name uniqueness in TypeScript");
  results.uniqueness = scoreUniqueness(nameOne, nameTwo, frequencyData);

  // Remove nicknames before the actual comparison
  console.error("Removing nicknames in TypeScript");
  [nameOne, nameTwo] = removeNicknames(nameOne, nameTwo);

  // 1st attempt: Checks if names are a match according to string comparison alone
  console.error("Starting TypeScript attempt one");
  var [attemptOneMatch, attemptOneWordCombos, attemptOneScore] = compareSpelling(nameOne, nameTwo);
  attemptOneScore = attemptOneScore + appliedPenalty;
  results.attemptOne = new Attempt(nameOne, nameTwo, attemptOneWordCombos, attemptOneScore);
  if (attemptOneMatch) {
    results.match = true;
    results.mostRecentAttemptScore = attemptOneScore;
    results.averageScoreOfCombinedAttempts = attemptOneScore;
    return results;
  };

  console.error(`Check to make sure the names weren't modified strangely after attempt one in TypeScript: nameOne - ${nameOne} nameTwo - ${nameTwo}`);

  // Failed first attempt. Check if names are even worth continuing
  if (isWorthContinuing(nameOne, nameTwo) === false){
    results.notWorthContinuing = true;
    results.mostRecentAttemptScore = attemptOneScore;
    results.averageScoreOfCombinedAttempts = attemptOneScore;
    return results;
  };

  console.error(`Check to make sure the names weren't modified strangely after continuation check in TypeScript: nameOne - ${nameOne} nameTwo - ${nameTwo}`);

  // 2nd attempt: Modify names via spelling rules, then check again if match according to string comparison
  console.error("Starting TypeScript attempt two");
  const [modifiedNameOne, modifiedNameTwo] = modifyNamesTogether(nameOne, nameTwo);
  var [attemptTwoMatch, attemptTwoWordCombos, attemptTwoScore] = compareSpelling(modifiedNameOne, modifiedNameTwo);
  attemptTwoScore = attemptTwoScore + appliedPenalty;
  results.attemptTwo = new Attempt(modifiedNameOne, modifiedNameTwo, attemptTwoWordCombos, attemptTwoScore);
  if (attemptTwoMatch) {
    results.match = true;
    results.mostRecentAttemptScore = attemptTwoScore;
    results.averageScoreOfCombinedAttempts = ((attemptTwoScore + attemptOneScore) / 2);
    return results;
  };

  // This is just to help with a test case
  let testModifyOne = getIpa(modifiedNameOne);
  let testModifyTwo = getIpa(modifiedNameTwo);
  console.error(`Retrieved ipas in TypeScript: nameOne - ${testModifyOne} nameTwo - ${testModifyTwo}`);
  
  // 3rd attempt: Checks if modified names are a match according to pronunciation
  console.error("Starting TypeScript attempt three");
  let ipaOfModifiedNameOne = cleanIpa(getIpa(modifiedNameOne));
  let ipaOfModifiedNameTwo = cleanIpa(getIpa(modifiedNameTwo));
  console.error(`TypeScript cleaned ipas in attempt three: nameOne - ${ipaOfModifiedNameOne}, nameTwo - ${ipaOfModifiedNameTwo}`);
  [ipaOfModifiedNameOne, ipaOfModifiedNameTwo] = modifyIpasByComparison(ipaOfModifiedNameOne, ipaOfModifiedNameTwo);
  console.error(`TypeScript ipas after modifying by comparison in attempt three: nameOne - ${ipaOfModifiedNameOne}, nameTwo - ${ipaOfModifiedNameTwo}`);
  var [attemptThreeMatch, attemptThreeWordCombos, attemptThreeScore] = pronunciationComparison(ipaOfModifiedNameOne, ipaOfModifiedNameTwo, modifiedNameOne, modifiedNameTwo);
  attemptThreeScore = attemptThreeScore + appliedPenalty;
  results.attemptThree = new Attempt(ipaOfModifiedNameOne, ipaOfModifiedNameTwo, attemptThreeWordCombos, attemptThreeScore);
  if (attemptThreeMatch) {
    results.match = true;
    results.mostRecentAttemptScore = attemptThreeScore;
    results.averageScoreOfCombinedAttempts = ((attemptThreeScore + attemptTwoScore + attemptOneScore) / 3);
    return results;
  };

  // 4th attempt: Check if original names are a match according to pronunciation
  console.error("Starting TypeScript attempt four");
  let ipaOfNameOne = cleanIpa(getIpa(nameOne));
  let ipaOfNameTwo = cleanIpa(getIpa(nameTwo));
  console.error(`TypeScript cleaned ipas in attempt four: nameOne - ${ipaOfNameOne}, nameTwo - ${ipaOfNameTwo}`);
  [ipaOfNameOne, ipaOfNameTwo] = modifyIpasByComparison(ipaOfNameOne, ipaOfNameTwo);
  console.error(`TypeScript ipas after modifying by comparison in attempt four: nameOne - ${ipaOfNameOne}, nameTwo - ${ipaOfNameTwo}`);
  var [attemptFourMatch, attemptFourWordCombos, attemptFourScore] = pronunciationComparison(ipaOfNameOne, ipaOfNameTwo, nameOne, nameTwo);
  attemptFourScore = attemptFourScore + appliedPenalty;
  results.attemptFour = new Attempt(ipaOfNameOne, ipaOfNameTwo, attemptFourWordCombos, attemptFourScore);

  if (attemptFourMatch) {
    results.match = true;
  };

  // Since we want the return values to be right, we need this to be a default case for them
  results.mostRecentAttemptScore = attemptFourScore;
  results.averageScoreOfCombinedAttempts = ((attemptFourScore + attemptThreeScore + attemptTwoScore + attemptOneScore) / 4);

  return results;
};
